metaseq_cli/train.py

```diff
@@ -486,6 +500,43 @@ def post_checkpoint_callback(cfg, filename):
             except (FileNotFoundError, AssertionError) as e:
                 logger.info(f"could not upload {filename}: {e}")
 
+        if do_evaluate:
+            _run_evaluations(
+                cfg.checkpoint.eval_command,
+                cfg.checkpoint.cloud_upload_path,
+                filename,
+                **eval_kwargs,
+            )
+
+
+def _run_evaluations(eval_cmd, cloud_upload_path, local_file, checkpoint_suffix, gloo_pg):
+    # Make sure all ranks have finished uploading checkpoints.
+    # If any rank doesn't hit the barrier within the timeout period, we throw an error and do
+    # not run evals. Error doesn't stop training run.
+    dist.monitored_barrier(group=gloo_pg, timeout=timedelta(minutes=5))
+    # Run evals on rank 0
+    if distributed_utils.get_global_rank() != 0:
+        return
+    assert eval_cmd is not None, "--eval-command needs to be set."
+    cmd = eval_cmd.split()
+    checkpoint_name = local_file.split("/")[-1].replace(checkpoint_suffix, "")
+    # The evals command must accept these two args and properly merge the cloud path
+    # with the checkpoint name.
+    cmd.extend(
+        [
+            "--model-cloud-path",
```

Comment:
@ruanslv could you provide an example of what the cloud_upload_path and  checkpoint_name fields look like?

For context, on evals side, right now we're reusing the logic to specify a local model path like
```
FSD=/data/gpt-z/few_shot_data/  python metaseq_internal/eval/gpt3_eval.py \
--model-name 125m \
--tasks copa --nshot 0 --fsdp \
--results-dir ./local_results/testing \
--distributed-world-size 2 \
--model-path az://lrsstoragewest3/data/opt/gptzmodels/gptz/125M/reshard_no_os/reshard.pt \
--model-template gptz_sharded_config
```

Reply:
Added a logger line for us to know which command is being run. Example:

`2022-10-26 20:58:11 | INFO | metaseq_cli.train | Kicking off evaluation cmd: ['python', 'metaseq_internal/scripts/eval/schedule_jobs_few_shot_opt_evaluation.py', '-o', './zucchini_evals/hookv', '-t', 'opt_eval_tasks', '--model-name', 'hookv', '--model-template', 'bf16_sharded_config', '--slurm-partition', 'zetta', '--model-path', '/data/gpt-z/models/zucchini_ablations/swiglu_fat/checkpoint_last.pt', '--model-cloud-path', 'https://lrsstoragewest3.blob.core.windows.net/ruanrms/checkpoints/hookv/?sv=2020-08-04&ss=bfqt&srt=sco&sp=rwdlacupitfx&se=2023-10-06T21:33:38Z&st=2022-05-09T13:33:38Z&spr=https&sig=lrvUOVNoS4p62JY3EV4XnNuJwq517ChOdFwitywQgeA%3D', '--model-cloud-filename', 'checkpoint_30.pt']`

Reply:
Compared to your command, basically everything is set in the training script, apart from `--model-path` which we need to infer based on `--model-cloud-path` and `--model-cloud-filename`. 

It's not a simple concatenation given we still use complete Azure paths for training, which have the auth stuff at the end. Otherwise we could construct the full model-path here too

Reply:
Discussed offline about how it might be better to turn this into just a python function we import and call, and we start version controlling our eval stuff inside metaseq-internal.